include/wil/network.h

```diff
@@ -0,0 +1,1338 @@
+        using difference_type = size_t;
+        using distance_type = size_t;
+        using pointer = ::wil::network::socket_address*;
+        using reference = ::wil::network::socket_address&;
+
+        constexpr addr_info_iterator_t() WI_NOEXCEPT = default;
+
+        explicit addr_info_iterator_t(const T* addrinfo) WI_NOEXCEPT :
+            // must const cast so we can re-use this pointer as we walk the list
+            m_addrinfo_ptr(const_cast<T*>(addrinfo))
+        {
+            if (m_addrinfo_ptr)
+            {
+                m_socket_address.set_sockaddr(m_addrinfo_ptr->ai_addr, m_addrinfo_ptr->ai_addrlen);
+            }
+        }
+
+        ~addr_info_iterator_t() WI_NOEXCEPT = default;
+        addr_info_iterator_t(const addr_info_iterator_t&) WI_NOEXCEPT = default;
+        addr_info_iterator_t& operator=(const addr_info_iterator_t&) WI_NOEXCEPT = default;
+        addr_info_iterator_t(addr_info_iterator_t&&) WI_NOEXCEPT = default;
+        addr_info_iterator_t& operator=(addr_info_iterator_t&&) WI_NOEXCEPT = default;
+
+        const ::wil::network::socket_address& operator*() const WI_NOEXCEPT
+        {
+            return m_socket_address;
+        }
+
+        const ::wil::network::socket_address& operator*() WI_NOEXCEPT
+        {
+            return m_socket_address;
+        }
+
+        const ::wil::network::socket_address* operator->() const WI_NOEXCEPT
+        {
+            return &m_socket_address;
+        }
+
+        const ::wil::network::socket_address* operator->() WI_NOEXCEPT
+        {
+            return &m_socket_address;
+        }
+
+        [[nodiscard]] bool operator==(const addr_info_iterator_t& rhs) const WI_NOEXCEPT
+        {
+            return m_addrinfo_ptr == rhs.m_addrinfo_ptr;
+        }
+
+        [[nodiscard]] bool operator!=(const addr_info_iterator_t& rhs) const WI_NOEXCEPT
+        {
+            return !(*this == rhs);
+        }
+
+        addr_info_iterator_t& operator++() WI_NOEXCEPT
+        {
+            this->operator+=(1);
+            return *this;
+        }
+
+        addr_info_iterator_t operator++(int) WI_NOEXCEPT
+        {
+            auto return_value = *this;
+            this->operator+=(1);
+            return return_value;
+        }
+
+        addr_info_iterator_t& operator+=(size_t offset) WI_NOEXCEPT
+        {
+            for (size_t count = 0; count < offset; ++count)
+            {
+                WI_ASSERT(m_addrinfo_ptr);
+                if (m_addrinfo_ptr)
+                {
+                    m_addrinfo_ptr = m_addrinfo_ptr->ai_next;
+
+                    if (m_addrinfo_ptr)
+                    {
+                        m_socket_address.set_sockaddr(m_addrinfo_ptr->ai_addr, m_addrinfo_ptr->ai_addrlen);
+                    }
+                    else
+                    {
+                        m_socket_address.reset();
+                    }
+                }
+            }
+
+            return *this;
+        }
+

```

Comment:
```suggestion

```
This is a forward iterator. It should not have `operator+=`.

The fact that this is a `for` loop - and therefore not O(1) - can be surprising and unexpected.